packages/zksync-dai-bridge/contracts/l2/dai.sol

```diff
@@ -0,0 +1,274 @@
+        uint256 allowed = allowance[msg.sender][spender];
+        require(allowed >= subtractedValue, "Dai/insufficient-allowance");
+        unchecked{
+            allowed = allowed - subtractedValue;
+        }
+        allowance[msg.sender][spender] = allowed;
+
+        emit Approval(msg.sender, spender, allowed);
+
+        return true;
+    }
+
+    // --- Mint/Burn ---
+    function mint(address to, uint256 value) external auth {
+        require(to != address(0) && to != address(this), "Dai/invalid-address");
+        unchecked {
+            balanceOf[to] = balanceOf[to] + value; // note: we don't need an overflow check here b/c balanceOf[to] <= totalSupply and there is an overflow check below
+        }
+        totalSupply = totalSupply + value;
+
+        emit Transfer(address(0), to, value);
+    }
+
+    function burn(address from, uint256 value) external {
+        uint256 balance = balanceOf[from];
+        require(balance >= value, "Dai/insufficient-balance");
+
+        if (from != msg.sender) {
+            uint256 allowed = allowance[from][msg.sender];
+            if (allowed != type(uint256).max) {
+                require(allowed >= value, "Dai/insufficient-allowance");
+
+                unchecked {
+                    allowance[from][msg.sender] = allowed - value;
+                }
+            }
+        }
+
+        unchecked {
+            balanceOf[from] = balance - value; // note: we don't need overflow checks b/c require(balance >= value) and balance <= totalSupply
+            totalSupply     = totalSupply - value;
+        }
+
+        emit Transfer(from, address(0), value);
+    }
+
+    // --- Approve by signature ---
+    
+    function _isValidSignature(
+        address signer,
+        bytes32 digest,
+        bytes memory signature
+    ) internal view returns (bool) {
+        if (signature.length == 65) {
+            bytes32 r;
+            bytes32 s;
+            uint8 v;
+            assembly {
+                r := mload(add(signature, 0x20))
+                s := mload(add(signature, 0x40))
+                v := byte(0, mload(add(signature, 0x60)))
+            }
+            if (signer == ecrecover(digest, v, r, s)) {
+                return true;
+            }
+        }
+
+        (bool success, bytes memory result) = signer.staticcall(
+            abi.encodeWithSelector(IERC1271.isValidSignature.selector, digest, signature)
+        );
+        return (success &&
+            result.length == 32 &&
+            abi.decode(result, (bytes4)) == IERC1271.isValidSignature.selector);
+    }
+
+    function permit(
+        address owner,
+        address spender,
+        uint256 value,
+        uint256 deadline,
+        bytes memory signature
+    ) public {
+        require(block.timestamp <= deadline, "Dai/permit-expired");
+        require(owner != address(0), "Dai/invalid-owner");
+
+        uint256 nonce;
+        unchecked { nonce = nonces[owner]++; }
```

Comment:
So this is something I introduced probably without realizing, comparing to the Optimism version, the `nonce` is starting from 1 instead of 0. Are we ok with it, or should just move the ++ after the digest calculation?
Or maybe just keeping the nonce[owner]++ with safe math in the digest. Or another option is putting the whole digest inside the unchecked block.

Reply:
Deleting this as it is wrong.